Sources/GRPC/GRPCKeepaliveHandlers.swift

```diff
@@ -0,0 +1,354 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import NIO
+import NIOHTTP2
+
+/// Provides keepalive pings.
+///
+/// The logic is determined by the gRPC keepalive
+/// [documentation] (https://github.com/grpc/grpc/blob/master/doc/keepalive.md).
+internal class GRPCClientKeepaliveHandler: ChannelInboundHandler, _ChannelKeepaliveHandler {
+  typealias InboundIn = HTTP2Frame
+  typealias OutboundOut = HTTP2Frame
+
+  init(configuration: ClientConnectionKeepalive, connectionManager: ConnectionManager) {
+    self.connectionManager = connectionManager
+    self.pingHandler = PingHandler(
+      pingCode: 5,
+      interval: configuration.interval,
+      timeout: configuration.timeout,
+      permitWithoutCalls: configuration.permitWithoutCalls,
+      maximumPingsWithoutData: configuration.maximumPingsWithoutData,
+      minimumSentPingIntervalWithoutData: configuration.minimumSentPingIntervalWithoutData
+    )
+  }
+
+  /// The connection manager.
+  let connectionManager: ConnectionManager?
+
+  /// The ping handler.
+  let pingHandler: PingHandler
+
+  /// The scheduled task which will ping.
+  var scheduledPing: Scheduled<Void>? = nil
+
+  /// The scheduled task which will close the connection.
+  var scheduledClose: Scheduled<Void>? = nil
+
+  func close(context: ChannelHandlerContext) {
+    self.connectionManager?.idle()
+    context.close(mode: .all, promise: nil)
+  }
+}
+
+internal class GRPCServerKeepaliveHandler: ChannelInboundHandler, _ChannelKeepaliveHandler {
+  typealias InboundIn = HTTP2Frame
+  typealias OutboundOut = HTTP2Frame
+
+  init(configuration: ServerConnectionKeepalive) {
+    self.pingHandler = PingHandler(
+      pingCode: 10,
+      interval: configuration.interval,
+      timeout: configuration.timeout,
+      permitWithoutCalls: configuration.permitWithoutCalls,
+      maximumPingsWithoutData: configuration.maximumPingsWithoutData,
+      minimumSentPingIntervalWithoutData: configuration.minimumSentPingIntervalWithoutData,
+      minimumReceivedPingIntervalWithoutData: configuration.minimumReceivedPingIntervalWithoutData,
+      maximumPingStrikes: configuration.maximumPingStrikes
+    )
+  }
+
+  /// The ping handler.
+  let pingHandler: PingHandler
+
+  /// The scheduled task which will ping.
+  var scheduledPing: Scheduled<Void>? = nil
+
+  /// The scheduled task which will close the connection.
+  var scheduledClose: Scheduled<Void>? = nil
+
+  func close(context: ChannelHandlerContext) {
+    context.close(mode: .all, promise: nil)
+  }
+}
+
+protocol _ChannelKeepaliveHandler: ChannelInboundHandler {
+  var pingHandler: PingHandler { get }
+  var scheduledPing: Scheduled<Void>? { get set }
+  var scheduledClose: Scheduled<Void>? { get set }
+  func close(context: ChannelHandlerContext)
+}
+
+extension _ChannelKeepaliveHandler where OutboundOut == HTTP2Frame, InboundIn == HTTP2Frame {
```

Comment:
Rather than put these constraints on the extension, let's require these constraints as part of the protocol.